docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -294,254 +296,285 @@ TODO
+  lanes first, preserving intra-lane FIFO ordering.
+- [[Transaction dissemination](#transaction-dissemination)] We will continue to use the current P2P channel for
+   disseminating transactions, and we will implement in the mempool the logic for selecting the
+   order in which to send transactions.
+
+### Lanes definition
+
+The list of lanes and their corresponding priorities will be hardcoded in the application logic. A
+priority is a value of type `uint32`, with 0 being a reserved value (see below). The application
+also needs to define which of the lanes it defines is the default lane. 
+
+To obtain the lane information from the application, we need to extend the ABCI `Info` response to
+include the following fields. These fields need to be filled by the application only in case it
+wants to implement lanes.
+```protobuf
+message InfoResponse {
+  ...
+  repeated uint32 lanes = 6;
+  uint32 default_lane = 7;
+}
+```
+Internally, the application may use `string`s to name lanes, and then map those names to priorities.
+The mempool does not care about the names, only about the priorities. That is why the lane
+information returned by the application only contains priorities.
+
+The lowest priority a lane may have is 1. The value 0 is reserved for two cases: when there are no
+priorities and for `CheckTx` responses of invalid transactions.
+
+On receiving the information from the app, CometBFT will validate that:
+- `lanes` has no duplicates (values in `lanes` don't need to be sorted),
+- `default_value` is in `lanes` (the default lane is not necessarily the lane with the lowest
+  priority), and
+- the list `lanes` is empty if and only if `default_value` is 0.
+
+### Initialization
+
+Currently, querying the app for `Info` happens during the handshake between CometBFT and the app,
+during the node initialization, and only when state sync is not enabled. The `Handshaker` first
+sends an `Info` request to fetch the app information, and then replays any stored block needed to
+sync CometBFT with the app. The lane information is needed regardless of whether state sync is
+enabled, so one option is to query the app information outside of the `Handshaker`.
+
+In this proposed approach, updating the lane definitions will require a single governance proposal
+for updating the software. CometBFT will not need to deal with dynamic lane changes: it will just
+need to set up the lanes when starting up (whether afresh or in recovery mode).
+
+Different nodes also need to agree on the lanes they use. When a node connects to a peer, they both
+perform a handshake to agree on some basic information (see `DefaultNodeInfo`), including the
+version of the application they are executing. Since the application includes the lane definitions,
+both nodes will also agree on the lanes they implement. Therefore we don't need to modify the P2P
+handshake.
+
```

Comment:
We could add a paragraph here to explain compatibility with CosmoVisor-like approaches.

Reply:
Yes. That's important and I forgot.

Reply:
Will add a paragraph

Reply:
Working on it now...

Reply:
Done. Please take a look

Reply:
Thanks!